hummingbot/hummingbot
# fix/user balance failure for time synchronizer

**Before submitting this PR, please make sure**:


**A description of the changes proposed in the pull request**:
Change in the logic that gets the user balance for Binance and Binance Perpetual. Since the exchange requires the time to be synchronized before any interaction, with this PR we add logic to the _update_balances() function to ensure the TimeSynchronizer has been initialized before requesting the balance.


**Tests performed by the developer**:
Added unit tests.


**Test summary**:
- Install the feature branch and start the Hummingbot client without any issues
- Adjust the machine time ahead by 10 minutes or more and connect the Binance and Binance perpetual API keys successfully
- Checked the balance and confirmed that balances are displayed
- Adjust the machine time by delaying it for about 10 minutes or more and rechecking the API key connections and confirming that keys are still confirmed.
- Started perpetual market-making and pure market-making strategies using the two connectors without any more issues or errors.
- Open the Binance exchange and check the timestamps of the open orders and confirmed that orders are using the Binance server time
- After an order gets filled or a position is opened and closed, checked the timestamp in the order history or trade history and confirmed that they used the Binance server time.


**Tips for QA testing**:
Run the balance command in a client instance with Binance and Binance Perpetual connectors correctly configured, and the machine time delayed 10 minutes.
Please consider that Binance and Binance Perpetual connectors can still sometimes cause an error when trying to process trade fills with fees in BNB. This error will keep happening until this PR is merged: https://github.com/hummingbot/hummingbot/pull/5160


This PR comes from https://github.com/CoinAlpha/hummingbot/pull/107
